src/libraries/CalldataDecoder.sol

```
@@ -11,35 +11,61 @@ library CalldataDecoder {
     function decodeActionsRouterParams(bytes calldata _bytes)
         internal
         pure
         returns (bytes calldata actions, bytes[] calldata params)
     {
         assembly ("memory-safe") {
-            // The offset of the 0th element is 0, which stores the offset of the length pointer of actions array.
-            // The offset of the 1st element is 32, which stores the offset of the length pointer of params array.
-            let actionsPtr := add(_bytes.offset, calldataload(_bytes.offset))
-            let paramsPtr := add(_bytes.offset, calldataload(add(_bytes.offset, 0x20)))
+            // Strict encoding requires that the data begin with:
+            // 0x00: 0x40 (offset to `actions.length`)
+            // 0x20: 0x60 + actions.length (offset to `params.length`)
+            // 0x40: `actions.length`
+            // 0x60: beginning of actions
+
+            // Verify actions offset matches strict encoding
+            let invalidData := xor(calldataload(_bytes.offset), 0x40)
+            actions.offset := add(_bytes.offset, 0x60)
+            actions.length := and(calldataload(add(_bytes.offset, 0x40)), OFFSET_OR_LENGTH_MASK)
 
-            // The length is stored as the first element
-            actions.length := calldataload(actionsPtr)
-            params.length := calldataload(paramsPtr)
+            // Round actions length up to be word-aligned, and add 0x60 (for the first 3 words of encoding)
+            let paramsLengthOffset := add(and(add(actions.length, 0x1f), OFFSET_OR_LENGTH_MASK_AND_WORD_ALIGN), 0x60)
+            // Verify params offset matches strict encoding
+            invalidData := or(invalidData, xor(calldataload(add(_bytes.offset, 0x20)), paramsLengthOffset))
+            let paramsLengthPointer := add(_bytes.offset, paramsLengthOffset)
+            params.length := and(calldataload(paramsLengthPointer), OFFSET_OR_LENGTH_MASK)
```

Here the `params.length` value is silently trimmed to 32 bytes, which could hide errors in abi encoding.  Consider doing instead:

    params.length := calldataload(paramsLengthPointer)
    invalidData := or(invalidData, shr(32, actions.length))